Add initial support for publishing prometheus metrics

dgoodwin opened the pull request:
This is just a starter PR, @ShivamChamoli will be taking over writing most of these. Vendors prometheus, exposes a couple cluster deployment controller metrics, and establishes a metrics calc loop outside of controllers for more global totals calculated every 2 minutes.

This can be tested with "make run" and "curl http://localhost:2112/metrics".

csrwng commented on cmd/manager/main.go:
```diff
@@ -139,7 +144,29 @@ func (writer glogWriter) Write(data []byte) (n int, err error) {
+		Client:   metricsClient,
+		Interval: 2 * time.Minute,
+	}
+	go mc.Run()
```
I would add this as a runnable to the manager

csrwng replied:
and add a separate add_metrics.go file for it

csrwng commented on cmd/manager/main.go:
```diff
@@ -139,7 +144,29 @@ func (writer glogWriter) Write(data []byte) (n int, err error) {
+	go http.ListenAndServe(":2112", nil)
+	log.Info("started /metrics http handler on port 2112")
+
+	metricsScheme := scheme.Scheme
```
From the variable, you get the impression that you're using a scheme just for metrics, but it's really an alias for the global scheme.Scheme. It may be better just to use that without the alias so it's understood that you're changing the global one.

dgoodwin replied:
Moving to the manager start actually let me use manager client and scheme so that cleans this up even more.

dgoodwin commented:
Updated to use controller-runtime metrics registry, removed my http listener, removed my obsolete metrics in cluster deployment controller, and now launching our metrics goroutine via the manager (and using it's client+scheme)

dgoodwin commented on pkg/controller/metrics/metrics.go:
```diff
@@ -0,0 +1,103 @@
+func init() {
+	metrics.Registry.MustRegister(metricClusterDeploymentsTotal)
+	metrics.Registry.MustRegister(metricClusterDeploymentsInstalledTotal)
+}
```
@ShivamChamoli just as a note for you, the above is how you can create and register a metric in the cluster deployment controller.

dgoodwin commented on pkg/controller/utils/utils.go:
```diff
@@ -92,3 +93,20 @@ func DeleteFinalizer(object metav1.Object, finalizer string) {
 }
+
+// GetKubeClient creates a new Kubernetes dynamic client.
+func GetKubeClient(scheme *runtime.Scheme) (client.Client, error) {
```
@csrwng if it's ok i'd propose keeping this function, it could be generally useful.

csrwng commented:
/lgtm
